File: 2023/day-09/src/bin/part_2.rs
fn main() {
    let input = include_str!("../input.txt");
    let output = solution(input);
    println!("Solution:\n{output}");
}

fn solution(input: &str) -> i32 {
    let mut total = 0;

    for line in input.lines() {
        let values = line
            .split_whitespace()
            .map(|x| x.parse::<i32>().unwrap())
            .collect::<Vec<i32>>();

        let mut cur_values: Vec<i32> = values.clone();

        let mut first_values: Vec<i32> = Vec::new();

        first_values.push(*cur_values.first().unwrap());

        while !cur_values.iter().all(|&x| x == 0) {
            let mut temp: Vec<i32> = Vec::new();
            for i in 0..(cur_values.len() - 1) {
                temp.push(cur_values[i + 1] - cur_values[i]);
            }

            cur_values = temp.clone();

            first_values.push(*temp.first().unwrap());
        }

        let cur = first_values.iter().rev().fold(0, |acc, &num| num - acc);

        total += cur;
    }

    total
}

File: 2022/day-06/part_2.py
def main():
    input = open(0).read()
    total =0

    for line in input.splitlines():
        cur = {}
        start = 0
        for (i,ch) in enumerate(line):
            if i-start == 14:
                total += start+14
                break
            if ch in cur and start <= cur[ch]:
                start = cur[ch] + 1


            cur[ch] = i

    return total

if __name__ == '__main__':
    print(main())
